EEPROMKeymap redesign

algernon opened the pull request:
Instead of having a single `keymap.map` Focus command that tries to be smart about what layers it presents, we should have two separate commands: one to query the defaults (PROGMEM), and one to get/set the custom keymap (EEPROM). This makes `keymap.map` and `keymap.roLayers` obsolete, and they're now removed.

Furthermore, having to specify whether the EEPROM keymap extends the built-in one or not proved to be unflexible. So we re-purposed the highest bit of the first EEPROM byte, to signal whether we should use EEPROM layers only or not.

algernon commented:
This is a big one, and it's kind of a breaking change, because it breaks the existing Focus protocol by removing two commands. Chrysalis will have to be updated in lock-step.

However, I'm fairly confident that this is the right direction, because separating built-in (default) and custom keymaps makes many things so much simpler, and far less surprising. Not to mention: more efficient! With the default firmware, we have 192 keys less to send back to the keyboard on update, which translates to more than half a kilobyte not being sent over 9600 baud. This is noticeable.

It also makes it simple to copy the default keymap to EEPROM, even from the CLI: `focus keymap.custom $(focus keymap.default)`. Compare that to the abomination I have in my progmem-to-eeprom script: `focus "keymap.map $(focus keymap.map | sed -e "s#^\(\([0-9]* *\)\{384\}\).*#\1\1#")"`

I'll take the first one any time of day.

The split also makes it easier to reason about `keymap.useCustom`. We currently require the users to decide whether to extend their PROGMEM layers with EEPROM, or do something different, at compile time. For keyboardio/Chrysalis#228, we want to do this at run-time. `keymap.map` currently returns the effective keymap. If we continued doing that, we wouldn't have access to the PROGMEM layers if `useCustom` is set, so we'd have to break the behaviour *anyway*. With the split, there's no dissonance: we always have access to both.

obra commented on src/kaleidoscope/plugin/EEPROM-Settings.cpp:
```diff
@@ -54,21 +54,30 @@ uint8_t EEPROMSettings::default_layer(uint8_t layer) {
   }
-  if (layer == 0xff) {
-    settings_.default_layer = layer;
+  if (layer >= 0x7e) {
```
Maybe move 0x7e to a named constant and add a line of doc about what it means?

obra commented on src/kaleidoscope/plugin/EEPROM-Settings.cpp:
```diff
@@ -54,21 +54,30 @@ uint8_t EEPROMSettings::default_layer(uint8_t layer) {
+   * value (since the highest bit is not ours), 0x7e is the explicit "don't mess
+   * with the default layer" setting. In either of these cases, we skip moving
+   * to the default layer. */
+  if (settings_.default_layer < 0x7e)
```
Oh hey. you do explain it :) But extract the constant and put the docs on the constant :)

obra commented on doc/plugin/EEPROM-Settings.md:
```diff
@@ -64,8 +64,18 @@ The plugin provides the `EEPROMSettings` object, which has the following methods
-> above.
+> Setting it to `126` or anything higher disables the automatic switching.
+
+### `use_eeprom_layers_only([true|false])`
```
perhaps "ignoreHardcodedLayers"?

algernon replied:
`ignoreBuiltInLayers`? or `ignoreDefaultLayers`? We refer to them as default layers elsewhere - not sure if that's a good thing, but however we refer to them, adjusting this name to match would be best I think.

obra replied:
"Default" feels funny to me. I'm ok with "BuiltIn", but since I' usually spell it "Builtin", I'm not in love with it. IgnoreFirmwareLayers? IgnoreCompiledLayers?

algernon replied:
`IgnoreBuiltinLayers` works for me, not attached to capitalizing that `I`. `IgnoreCompiledLayers` would be my second choice, `IgnoreFirmwareLayers`.... makes me feel uneasy, not quite sure why.

obra replied:
(My worry about capitalization and that I is that it'll add cognitive load remembering which way we have it, rather than preferring any specific version of it.)

Hardcoded doesn't do it for you?

algernon replied:
It beats the name I gave the method, but it doesn't feel right to me. I don't have any better suggestions however, so we can go with `IgnoreHardcodedLayers`. It's not very important, because the primary user of this is `EEPROMKeymap`, and it presents it under a different name (`keymap.useCustom`). IOW, it's not user-facing in the majority of cases.

obra commented:
useCustom also doesn't -really- explain what we're doing. would onlyCustom
make sense?
ᐧ

On Tue, Jan 29, 2019 at 4:00 PM Gergely Nagy <notifications@github.com>
wrote:

> *@algernon* commented on this pull request.
> ------------------------------
>
> In doc/plugin/EEPROM-Settings.md
> <https://github.com/keyboardio/Kaleidoscope/pull/553#discussion_r252069228>
> :
>
> > @@ -64,8 +64,18 @@ The plugin provides the `EEPROMSettings` object, which has the following methods
>  > keyboard boots up, it will automatically switch to the configured layer - if
>  > any.
>  >
> -> This is the Focus counterpart of the `default_layer()` method documented
> -> above.
> +> Setting it to `126` or anything higher disables the automatic switching.
> +
> +### `use_eeprom_layers_only([true|false])`
>
> It beats the name I gave the method, but it doesn't feel right to me. I
> don't have any better suggestions however, so we can go with
> IgnoreHardcodedLayers. It's not very important, because the primary user
> of this is EEPROMKeymap, and it presents it under a different name (
> keymap.useCustom). IOW, it's not user-facing in the majority of cases.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/keyboardio/Kaleidoscope/pull/553#discussion_r252069228>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACxaLuHH1yVFTv7XpEqQxZ_k1k-9iCcks5vIOC3gaJpZM4aW8-x>
> .
>

algernon commented:
`onlyCustom` sounds good

obra commented:
Great. Happy to re-review at your leisure
ᐧ

On Tue, Jan 29, 2019 at 4:07 PM Gergely Nagy <notifications@github.com>
wrote:

> onlyCustom sounds good
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/keyboardio/Kaleidoscope/pull/553#issuecomment-458758469>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACxaJBhx6zm1NGE3P6A6EYOKnhYyQgTks5vIOI_gaJpZM4aW8-x>
> .
>

algernon commented:
Updated the PR, now it uses `ignoreHardcodedLayers()` and `keymap.onlyCustom`, and I introduced a constant instead of `0x7e`. Also made the check a bitmask check, which does essentially the same thing, but we end up saving 12 bytes of PROGMEM compared to the `>=` check.
